File: src/1-extending-ruby/leibniz/ext/leibniz/leibniz.rb
require 'ffi'

module Leibniz
  module C
    extend FFI::Library
    ffi_lib './leibniz.so'

    attach_function :calc, :normal, [:size_t], :double
  end

  module SIMD
    extend FFI::Library
    ffi_lib './leibniz.so'

    attach_function :calc, :simd, [:size_t], :double
  end

  module Ruby
    def self.calc(n)
      signal = -1.0
      pi = 0.0

      n.times do
        signal = -signal
        pi += signal / (2 * it + 1)
      end

      pi * 4.0
    end
  end
end

File: src/1-extending-ruby/leibniz/ext/leibniz/leibniz.c
#include <ruby.h>
#include <stdio.h>

VALUE calc(VALUE self, VALUE times) {
  size_t n = RB_NUM2SIZE(times);
  double pi = 0.0;
  double signal = -1.0;

  for(size_t i = 0; i < n; ++i) {
    signal = -signal;
    pi += signal / (2 * i + 1);
  }

  return rb_float_new(pi * 4.0);
}

void Init_leibniz(void) {
  VALUE leibnizModule = rb_define_module("Leibniz");
  rb_define_singleton_method(leibnizModule, "calc", calc, 1);
}
